### PR TITLE
Add support for highlighting enclosing brackets

### PR DESCRIPTION
**Pull request by riverbl**
Highlight enclosing brackets when `highlight_matching_brackets` is enabled.

Draw lines delineating the enclosing scope when `highlight_scope_lines` is enabled.

Currently only draws the highlights and scope lines for one cursor, even if there are multiple cursors.

Currently doesn't draw highlights or scope lines on the sticky header.

Relates to #2417.

- [ ] Added an entry to `CHANGELOG.md` if this change could be valuable to users

### CODE REVIEW

File: lapce-app/src/editor/view.rs [hunk 1]
```diff
@@ -843,6 +843,210 @@ impl EditorView {
...
+                .map(|line_info| line_info.y as f64 + line_height);
+            let end_line_y = screen_lines
+                .info
+                .get(&end_line)
+                .map(|line_info| line_info.y as f64 + line_height);
+
+            // We only need to draw anything if start_line is on or before the visible section and
+            // end_line is on or after the visible section.
+            let y0 = start_line_y.or_else(|| {
+                (start_line < *screen_lines.lines.first().unwrap())
```
Comment (riverbl) [inline lapce-app/src/editor/view.rs:940]: Is it valid to assume that `screen_lines.lines` is non-empty and sorted?

File: lapce-core/src/syntax/mod.rs [hunk 2]
```diff
@@ -841,11 +835,13 @@ impl Syntax {
             if start >= self.text.len() {
                 return None;
             }
-            let c = self.text.byte_at(start) as char;
-            if matching_pair_direction(c) == Some(true) {
-                let end = self.find_matching_pair(start)?;
-                if end >= offset {
-                    return Some((start, end));
+            if start < offset {
```
Comment (riverbl) [inline lapce-core/src/syntax/mod.rs:838]: Previously the the offset immediately before an opening bracket was considered enclosed (i.e.: `|(    )`). This was inconsistent with `WordCursor::find_enclosing_pair`, which is used for language unaware bracket search, so I've changed it to not consider that offset to be enclosed.

File: lapce-core/src/word.rs [hunk 3]
```diff
@@ -428,7 +428,7 @@ impl<'a> WordCursor<'a> {
             if matching_pair_direction(c) == Some(true) {
                 let opening_bracket_offset = self.inner.pos();
                 if let Some(closing_bracket_offset) = self.match_pairs() {
-                    if (opening_bracket_offset..closing_bracket_offset)
+                    if (opening_bracket_offset..=closing_bracket_offset)
```
Comment (riverbl) [inline lapce-core/src/word.rs:431]: This fixes a bug where the offset immediately before the closing bracket (i.e.: `(   |)`) was not considered enclosed.

File: lapce-app/src/editor/view.rs (hunk 1 above)
Comment (dzhou121) [reply lapce-app/src/editor/view.rs:940]: It's sorted but could be empty I think.

Comment (riverbl) [reply lapce-app/src/editor/view.rs]: I've changed it to not assume that `screen_lines.lines` is non-empty.

**Pull request merged**